### PR TITLE
adds macro for making bounds checks a no-op

### PR DESCRIPTION
**Pull request by SteveBronder**
## Summary

From issue #2420 the default in Stan now is to always perform bounds checks for single index access. This has a noticeable slowdown for users. This PR uses a `STAN_NO_RANGE_CHECKS` macro which if defined then makes the range and bounds checks into no-ops. It does this by defining another macro `STAN_NO_RANGE_CHECK_RETURN` which is equal to `return;` if `STAN_NO_RANGE_CHECKS` is defined.

At first I thought we could just use `NDEBUG` but it seems that is specifically to turn on and off asserts such as `eigen_assert()` or the standard `assert()`. Our bounds checks are sort of an assert, though I think it's safer just to have a new macro.

It's possible for the user to specify this in the makefile, or we could have an option for the compiler like `-fno-debug` which at the top of the stan model would define `STAN_NO_RANGE_CHECKS`

## Tests

tbh I'm not really sure how to test this since it changes behavior in the preprocessor. We could have a separate jenkins node that does the checks just for these errors? Since they are all no-ops it would be pretty fast

## Side Effects

Yes this whole PR is a side effect to turn off bounds checks.

## Release notes

Adds `STAN_NO_RANGE_CHECKS` macro which if defined turns off bounds and range checks

## Checklist

- [x] Math issue #2420

- [x] Copyright holder: Steve Bronder

    The copyright holder is typically you or your assignee, such as a university or company. By submitting this pull request, the copyright holder is agreeing to the license the submitted work under the following licenses:
      - Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)
      - Documentation: CC-BY 4.0 (https://creativecommons.org/licenses/by/4.0/)

- [x] the basic tests are passing

    - unit tests pass (to run, use: `./runTests.py test/unit`)
    - header checks pass, (`make test-headers`)
    - dependencies checks pass, (`make test-math-dependencies`)
    - docs build, (`make doxygen`)
    - code passes the built in [C++ standards](https://github.com/stan-dev/stan/wiki/Code-Quality) checks (`make cpplint`)

- [ ] the code is written in idiomatic C++ and changes are documented in the doxygen

- [x] the new changes are tested

### CODE REVIEW

**Comment by rok-cesnovar** [comment]
If we will be defining a separate macro, which I think is the right call here, might as well give it a more descriptive name? Most of Stan users are not C++ developers and NDEBUG does not mean much to them. Maybe STAN_NO_RANGE_CHECKS or just NO_RANGE_CHECKS.

**Comment by rok-cesnovar** [comment]
As for testing, I think easiest is a Github Action job that runs a list of tests. These standalone short tests are perfect for GHA as its free.

**Comment by SteveBronder** [comment]
> If we will be defining a separate macro, which I think is the right call here, might as well give it a more descriptive name? Most of Stan users are not C++ developers and NDEBUG does not mean much to them. Maybe STAN_NO_RANGE_CHECKS or just NO_RANGE_CHECKS.

I have to think about that. Messing with macro's is certainly an advanced user function. I use `STAN_NO_RANGE_CHECKS` for the `return;`, though I could check if it exists and then undef and def it 

> As for testing, I think easiest is a Github Action job that runs a list of tests. These standalone short tests are perfect for GHA as its free.

Nice! Yeah that makes total sense I'll add the tests and make a gh action

File: stan/math/prim/err/check_bounded.hpp
Comment (bbbales2) [inline stan/math/prim/err/check_bounded.hpp:77]: This is used to check that real values are in a certain range (like probabilities are bounded [0, 1]) as well as indexes in range.

It looks like this pull is only turning off range checks, but why not value checks too? I think you'd turn them off in similar situations for the same reasons.

**Review by bbbales2 — COMMENTED** [review]
Comment.

**Comment by t4c1** [comment]
> I use STAN_NO_RANGE_CHECKS for the return;, though I could check if it exists and then undef and def it

You should probably rename that to `STAN_NO_RANGE_CHECKS_RETURN` and define that if a user defines `STAN_NO_RANGE_CHECKS`.

File: stan/math/prim/meta/compiler_attributes.hpp
Comment (t4c1) [inline stan/math/prim/meta/compiler_attributes.hpp:21]: These changes seem unrelated to what this PR is doing.

Comment (SteveBronder) [reply stan/math/prim/meta/compiler_attributes.hpp:21]: This is a little bit of a fix, if someone is using either not gcc or clang then we are not guaranteed `__has_attribute` is defined. So to compile on those we want to make sure that macro is defined so that we can then check whether the compiler has the attributes `noinline` and `cold`

File: stan/math/prim/err/check_bounded.hpp
Comment (SteveBronder) [reply stan/math/prim/err/check_bounded.hpp:77]: Oh I forgot this was used for both. Umm I think @syclik had some opinions on this. Also pinging @wds15. I figure this is something users have to explicitly turn on so I'm more in favor of also not doing value checks

Comment (wds15) [reply stan/math/prim/err/check_bounded.hpp:77]: Value checks? Like what?

Comment (bbbales2) [reply stan/math/prim/err/check_bounded.hpp:77]: Not NaN, not infinite, positive definite, valid unit vector, valid simplex -- this sort of thing.

Comment (wds15) [reply stan/math/prim/err/check_bounded.hpp:77]: that's a different story which one should be able to turn off separately, I think. Something like STAN_FAST_MATH (a synonym for "not safe math, do it fast")

Comment (bbbales2) [reply stan/math/prim/err/check_bounded.hpp:77]: If we're allowing for segfaults, then I am also happy allowing for random math explosions. Even though these could happen in really weird, hard to debug ways, it's all better when the checks are flipped back on.

Comment (bbbales2) [reply stan/math/prim/err/check_bounded.hpp:77]: Well I guess, how is being unsafe on values practically different than being unsafe on indexing? A lot of the time even when we see these value check errors in practice we ignore them (scale parameter is zero, etc. etc.)

I suppose that these can happen occasionally even in a Stan model we are happy with, and we don't regularly expect to hit out of bounds indexing and stuff.

Comment (bbbales2) [reply stan/math/prim/err/check_bounded.hpp:77]: To not stall this unnecessarily, I think it's fine if we do:

1. The macro turns off some but not all index/value checks (this base set seems fine)
2. Only doing the actual index checks for now (`check_bounded` no longer optimized)
3. The macro turns off some but not all index/value checks and we expand it to turn off inf/nan checks
4. Some other variation on a theme

I like one macro better than to two because it would probably just get confusing to have two (for functions like this for instance) and I think the conditions for no index checks are similar to no value checks.

Easy enough to expand the set of functions optimized in production mode at a later time if we only want to do strict indices.

File: stan/math/prim/meta/compiler_attributes.hpp
Comment (t4c1) [reply stan/math/prim/meta/compiler_attributes.hpp:21]: Ok, but you could simplify this into:
```suggestion
#ifdef __has_attribute
#if __has_attribute(noinline) && __has_attribute(cold)
/**
 * Functions tagged with this attribute are not inlined and moved
 *  to a cold branch to tell the CPU to not attempt to pre-fetch
 *  the associated function.
 */
#define STAN_COLD_PATH __attribute__((noinline, cold))
#endif
#ifndef STAN_COLD_PATH
#define STAN_COLD_PATH
#endif
```

File: stan/math/prim/err/check_bounded.hpp
Comment (wds15) [reply stan/math/prim/err/check_bounded.hpp:77]: Hmmm... that may work... but what is wrong with two macros which do a more tailored thing themselves. To retain easy usability we can introduce things like

```c++
#ifdef NDEBUG
#define MACRO1
#define MACRO2
#endif
```

How does that sound?

Comment (SteveBronder) [reply stan/math/prim/err/check_bounded.hpp:77]: I like in the future having a `STAN_NDEBUG` which turns off multiple types of checks. Following @t4c1 I changed the name of this macro to `STAN_NO_RANGE_CHECKS` which defines `STAN_NO_RANGE_CHECKS_RETURN return`

Comment (SteveBronder) [reply stan/math/prim/err/check_bounded.hpp:77]: Are we also not going to remove size checks?

Comment (bbbales2) [reply stan/math/prim/err/check_bounded.hpp:77]: > but what is wrong with two macros which do a more tailored thing themselves

I don't see why one would be useful without another -- unsafe is unsafe and it's just hard to remember multiple things.

> not going to remove size checks?

Optimizing out size checks makes sense with this macro I think. `check_bounded` is probly the only one that needs to go if we're not using values.

If we're going for STAN_NDEBUG in the future, does it make sense to just start with that? Or STAN_UNSAFE or STAN_NO_DEBUG since Rok pointed out NDEBUG probably doesn't mean much to people? Then in the future if we decide to split that into multiple things we can and the overall super-unsafe easy-mode flag is still there.

Comment (SteveBronder) [reply stan/math/prim/err/check_bounded.hpp]: I think I'd like to leave the size checks for another time because it changes the exceptions that are returned for some functions like `check_symmetric()` etc.

Comment (SteveBronder) [reply stan/math/prim/err/check_bounded.hpp]: > If we're going for STAN_NDEBUG in the future, does it make sense to just start with that? Or STAN_UNSAFE or STAN_NO_DEBUG since Rok pointed out NDEBUG probably doesn't mean much to people? Then in the future if we decide to split that into multiple things we can and the overall super-unsafe easy-mode flag is still there.

Let's do this in a separate PR once we sort out what that means. Right now I like just having an option to turn off the range checks, then in another PR or an issue we can sort out what we want turned on and off

Comment (bbbales2) [reply stan/math/prim/err/check_bounded.hpp]: Sounds good

File: stan/math/prim/meta/compiler_attributes.hpp
Comment (bbbales2) [inline stan/math/prim/meta/compiler_attributes.hpp:5]: Is it that old gccs didn't have likely and new ones do?

**Review by bbbales2 — APPROVED** [review]
This all looks good to me. Since we've been doing this pull by committee, I'll leave it to someone else to also look over and click the merge button.

File: stan/math/prim/meta/compiler_attributes.hpp
Comment (SteveBronder) [reply stan/math/prim/meta/compiler_attributes.hpp:5]: Another package could have it

**Pull request merged**
